### PR TITLE
feat: Action to download and verify an artifact by hash

### PR DESCRIPTION
**Pull request by laurentsimon**
depends on https://github.com/slsa-framework/slsa-github-generator/pull/436 which needs to be merged first

This is part of https://github.com/slsa-framework/slsa-github-generator/issues/26

This PR defines an Action to download and verify an artifact by hash. We can later replace the script with a better programming language.

### CODE REVIEW

File: .github/actions/secure-download-artifact/action.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    - name: Verify the SHA256
+      env:
+        UNTRUSTED_EXPECTED_HASH: "${{ inputs.sha256 }}"
+        UNTRUSTED_COMPUTED_HASH: "${{ steps.compute.outputs.sha256 }}"
+        UNTRUSTED_PATH: "${{ inputs.name }}"
+        UNTRUSTED_DST_PATH: "${{ inputs.path }}"
+        SET_EXECUTABLE: "${{ set-executable }}"
+      run: |
+        set -euo pipefail
```
Comment (ianlewis) [inline .github/actions/secure-download-artifact/action.yml:39]: as with the other PR, we can remove `set -euo pipefail` in this case.

File: .github/actions/secure-download-artifact/action.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+description: 'Download an artifact and verify its SHA256'
+inputs:
+  name:
+    description: 'Artifact name.'
+    required: true
+  path:
+    description: 'Destination path.'
+    required: false
+  sha256:
+    description: 'SHA256 of the file.'
```
Comment (ianlewis) [inline .github/actions/secure-download-artifact/action.yml:11]:
```suggestion
    description: 'SHA256 of the file for verification.'
```

**Review by ianlewis — APPROVED** [review]

File: .github/actions/secure-download-artifact/action.yml (hunk 1 above)
Comment (laurentsimon) [reply .github/actions/secure-download-artifact/action.yml:39]: thanks, done.

File: .github/actions/secure-download-artifact/action.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        UNTRUSTED_DST_PATH: "${{ inputs.path }}"
+        SET_EXECUTABLE: "${{ set-executable }}"
+      run: |
+        set -euo pipefail
+
+        ./.github/actions/secure-artifact-download/validate-sha256.sh
+
+
+     
+
```
[the remark this answers is not in the document]
Comment (laurentsimon) [reply .github/actions/secure-download-artifact/action.yml:45]: thanks, missed this during a copy/paste.

**Pull request merged**